web_exportall/controllers/main.py

```diff
@@ -0,0 +1,201 @@
+# -*- encoding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2012-2015 credativ Ltd (<http://credativ.co.uk>).
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as published
+#    by the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+import csv
+import operator
+import simplejson
+import re
+from cStringIO import StringIO
+try:
+    import xlwt
+except ImportError:
+    xlwt = None
+
+import openerp
+from openerp import sql_db
+from openerp.addons.web.controllers.main import (ExportFormat,
+                                                 serialize_exception)
+from openerp import http
+from openerp.http import request
+
+
+def itter_ids(all_ids, step_size=250):
+    """ Splits all_ids into more manageable chunks of
+        size step_size and yields them """
+    offset = 0
+    while True:
+        id_set = all_ids[offset:offset+step_size]
+        offset += step_size
+        if len(id_set):
+            yield id_set
+        else:
+            return
+
+
+def itter_data(Model, dbname, uid, all_ids, field_names, raw_data, context):
+    """ Reads data in chunks of ids from all_ids and
+        yields the rows one by one """
+    cr = sql_db.db_connect(dbname).cursor()
+    with cr:
+        with openerp.api.Environment.manage():
+            for ids in all_ids:
+                import_data = Model.export_data(cr,
+                                                uid,
+                                                ids,
+                                                field_names,
+                                                raw_data,
+                                                context=context
+                                                ).get('datas', [])
+                for data in import_data:
+                    yield data
+    return
+
+
+class ExportAllBase(ExportFormat):
```

Comment:
You can probably simplify the inheriting classes even more: from_data could be defined here with abstract functions overwitten in the inheriting classes for things such as `write_row`, `write_header`